man/fscrypt.8.md
fscrypt(8) -- manage linux filesystem encryption
================================================

## SYNOPSIS

**fscrypt** _command_ [arguments] [command options] [`--quiet` | `--verbose`]

**fscrypt** [_command_] `--help`

**fscrypt** `--version`

## DESCRIPTION

TODO

## WARNINGS

TODO

## ALTERNATIVE TOOLS

**fscrypt** only manages native filesystem encryption. The encryption tools
below may suit your needs better. 

**fscryptctl**(8) also manages filesystem encryption, but it does so through a
very low-level interface. It applies policy identifiers to directories, and
provisions keys into the kernel keyring. If you want to manage key derivation,
key rotation, metadata, and PAM integration yourself, this is a more lightweight
alternative.

Dm-crypt encrypts an entire block device with a single master key. dm-crypt can
be used with or without **fscrypt**. All filesystem data (including all
filesystem metadata) is encrypted with this single key when using dm-crypt,
while **fscrypt** only encrypts the filenames and file contents in a specified
directory. See **cryptsetup**(8) for more information.

It is possible to use both dm-crypt and **fscrypt** simultaneously, giving the
protections and benefits of both. One example of a reasonable setup could
involve using dm-crypt with a TPM or Secure boot key, while using **fscrypt**
on the contents of a home directory. This would still encrypt the entire drive,
but would also tie the encryption of a user's personal documents to their
passphrase. However, this may cause a decrease in your performance, as file
contents can be encrypted twice.  

eCryptfs is another form of filesystem encryption on Linux; it encrypts a
filesystem directory with some key or passphrase. eCryptfs sits on top of an
existing filesystem. This makes eCryptfs an alternative choice if your
filesystem or kernel does not support native filesystem encryption. See
**ecryptfs**(7) for more information.

## REQUIREMENTS

TODO

## OVERVIEW

TODO: Protectors, Policies, Keyring

## COMMANDS

**fscrypt** has multiple _command_ values, each of which can be used with the
common options (in this page) and command-specific options (found in the
below pages).

* **fscrypt-enable**(8):
    Enable encryption on an ext4 filesystem.
* **fscrypt-setup**(8):
    Create necessary global or per-filesystem files.
* **fscrypt-encrypt**(8):
    Start encrypting an empty directory.
* **fscrypt-unlock**(8):
    Unlock an encrypted directory.
* **fscrypt-purge**(8):
    Remove the keys for an encrypted directory.
* **fscrypt-status**(8):
    Print the status of the system, a filesystem, or a file.
* **fscrypt-metadata**(8):
    Manipulate the policy or protector metadata. **Warning:** this is an
    _expert_ command that can easily cause data loss. Use with care.

## OPTIONS

* `--help`:
    Show the help text for fscrypt, using the man pages if possible.
* `--version`:
    Show the version and copyright information.
* `--verbose`:
    Print additional debug messages to standard output.
* `--quiet`:
    Print nothing to standard output except for errors. Select the default for
    any options that would normally show a prompt.

## RETURN VALUES

On success, all **fscrypt** commands return 0. On failure, commands will return
1 and print the corresponding cause of failure to stderr.

## EXAMPLES

TODO

```bash
# Create the global configuration file. Nothing else needs root.
>>>>> sudo fscrypt setup
Create "/etc/fscrypt.conf"? [Y/n] y
Customizing passphrase hashing difficulty for this system...
Created global config file at "/etc/fscrypt.conf".
```

## BUGS

Any bugs, problems, or design discussion relating to **fscrypt** should be
raised in the
[Github Issue Tracker](https://github.com/google/fscrypt/issues/new).

**IMPORTANT:** Any significant security issues should **NOT** be reported in
the public issue tracker. Practice responsible disclosure by emailing
<joerichey@google.com> and <tyhicks@canonical.com> directly.

## AUTHOR

Joe Richey <joerichey94@gmail.com>

## COPYRIGHT

Copyright 2017 Google Inc. under the [Apache 2.0 License](https://www.apache.org/licenses/LICENSE-2.0).

## SEE ALSO

**fscrypt-enable**(8) **fscrypt-setup**(8) **fscrypt-encrypt**(8)
**fscrypt-unlock**(8) **fscrypt-purge**(8) **fscrypt-status**(8)
**fscrypt-metadata**(8) **fscrypt-config**(8) **pam_fscrypt**(8)
**fscryptctl**(8)

[**fscrypt**'s upstream repository](https://github.com/google/fscrypt) contains FAQs, known issues, longer examples, and information about building,
testing, and contributing to **fscrypt**.

cmd/fscrypt/fscrypt.go
/*
 * fscrypt.go - File which starts up and runs the application. Initializes
 * information about the application like the name, version, author, etc...
 *
 * Copyright 2017 Google Inc.
 * Author: Joe Richey (joerichey@google.com)
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not
 * use this file except in compliance with the License. You may obtain a copy of
 * the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
 * License for the specific language governing permissions and limitations under
 * the License.
 */

/*
fscrypt is a command line tool for managing linux filesystem encryption.
*/

// +build linux,cgo

package main

import (
	"fmt"
	"io/ioutil"
	"log"
	"os"
	"time"

	"github.com/urfave/cli"
)

var (
	// Current version of the program (set by Makefile)
	version string
	// Formatted build time (set by Makefile)
	buildTime string
	// Authors to display in the info command
	Authors = []cli.Author{{
		Name:  "Joe Richey",
		Email: "joerichey@google.com",
	}}
)

func main() {
	cli.AppHelpTemplate = appHelpTemplate
	cli.CommandHelpTemplate = commandHelpTemplate
	cli.SubcommandHelpTemplate = subcommandHelpTemplate

	// Create our command line application
	app := cli.NewApp()
	app.Usage = shortUsage
	app.Authors = Authors
	app.Copyright = apache2GoogleCopyright

	// Grab the version and compilation time passed in from the Makefile.
	app.Version = version
	app.Compiled, _ = time.Parse(time.UnixDate, buildTime)
	app.OnUsageError = onUsageError

	// Setup global flags
	cli.HelpFlag = helpFlag
	cli.VersionFlag = versionFlag
	cli.VersionPrinter = func(c *cli.Context) {
		cli.HelpPrinter(c.App.Writer, versionInfoTemplate, c.App)
	}
	app.Flags = universalFlags

	// We hide the help subcommand so that "fscrypt <command> --help" works
	// and "fscrypt <command> help" does not.
	app.HideHelp = true

	// Initialize command list and setup all of the commands.
	app.Action = defaultAction
	app.Commands = []cli.Command{Setup, Encrypt, Unlock, Purge, Status, Metadata}
	for i := range app.Commands {
		setupCommand(&app.Commands[i])
	}

	app.Run(os.Args)
}

// setupCommand performs some common setup for each command. This includes
// hiding the help, formating the description, adding in the necessary
// flags, setting up error handlers, etc... Note that the command is modified
// in place and its subcommands are also setup.
func setupCommand(command *cli.Command) {
	command.Description = wrapText(command.Description, indentLength)
	command.HideHelp = true
	command.Flags = append(command.Flags, universalFlags...)

	if command.Action == nil {
		command.Action = defaultAction
	}

	// Setup function handlers
	command.OnUsageError = onUsageError
	if len(command.Subcommands) == 0 {
		command.Before = setupBefore
	} else {
		// Cleanup subcommands (if applicable)
		for i := range command.Subcommands {
			setupCommand(&command.Subcommands[i])
		}
	}
}

// setupBefore makes sure our logs, errors, and output are going to the correct
// io.Writers and that we haven't over-specified our flags. We only print the
// logs when using verbose, and only print normal stuff when not using quiet.
// When running with sudo, this function also verifies that we have the proper
// keyring linkage enabled.
func setupBefore(c *cli.Context) error {
	log.SetOutput(ioutil.Discard)
	c.App.Writer = ioutil.Discard

	if verboseFlag.Value {
		log.SetOutput(os.Stdout)
	}
	if !quietFlag.Value {
		c.App.Writer = os.Stdout
	}
	return nil
}

// defaultAction will be run when no command is specified.
func defaultAction(c *cli.Context) error {
	// Always default to showing the help
	if helpFlag.Value {
		cli.ShowAppHelp(c)
		return nil
	}

	// Only exit when not calling with the help command
	var message string
	if args := c.Args(); args.Present() {
		message = fmt.Sprintf("command \"%s\" not found", args.First())
	} else {
		message = "no command was specified"
	}
	return &usageError{c, message}
}
